controller.py

```diff
@@ -0,0 +1,103 @@
+import typing
+
+import serial
+import enum
+import logging
+
+import util
+from state import State, Error
+
+
+LOGGER = logging.getLogger('controller')
+
+class StateKeys:
+    CURR_X = 'curr_x'
+    TRGT_X = 'trgt_x'
+    CURR_V = 'curr_v'
+    TRGT_V = 'trgt_v'
+    CURR_A = 'curr_a'
+    TRGT_A = 'trgt_a'
+    POLE_ANG = 'pole_ang'
+    POLE_VEL = 'pole_vel'
+    TIMESTAMP = 'timestamp'
+    ERRCODE = 'errcode'
+
+
+class ActionType(enum.Enum):
+    PositionControl = 1
+    VelocityControl = 2
+    AccelerationControl = 3
+
+
+class Action:
+    def __init__(self, action_type: ActionType, value: float) -> None:
+        self.type = action_type
+        self.value = value
+
+    def __str__(self) -> str:
+        action_str = {
+            ActionType.PositionControl: StateKeys.TRGT_X, 
+            ActionType.VelocityControl: StateKeys.TRGT_V,
+            ActionType.AccelerationControl: StateKeys.TRGT_A,
+            }[self.type]
+        return f'{action_str}={self.value}'
+
+
+class CartPoleController:
+    def __init__(self, device: str, baud_rate: int, step_n: int) -> None:
+        self.serial = serial.Serial(port=device, baudrate=baud_rate)
+        LOGGER.info(f'Opened serial connection to port {self.serial.name}')
+
+        self.step_n = step_n
+        self.step_count = 0
+
+    @staticmethod
+    def _parse_state(line: str) -> State:
+        kv = dict(kv_token.split('=') for kv_token in line.split())
+        return State(
+            cart_position=float(kv[StateKeys.CURR_X]),
+            cart_velocity=float(kv[StateKeys.CURR_V]),
+            pole_angle=float(kv[StateKeys.POLE_ANG]),
+            pole_velocity=float(kv[StateKeys.POLE_VEL]),
+            error=Error(int(kv[StateKeys.ERRCODE]))
+        )
+
+    @staticmethod
+    def _log_get_error(line: str):
+        if line.startswith('!'):
+            LOGGER.error(f'Failed to get state, got response: {line}')
+
+    def state(self) -> State:
+        self.serial.write(f'state get {StateKeys.CURR_X} {StateKeys.CURR_V} {StateKeys.CURR_A} '
+            f'{StateKeys.POLE_ANG} {StateKeys.POLE_VEL} {StateKeys.TIMESTAMP} {StateKeys.ERRCODE}')
+        line = self.serial.readline().lower()
+        return self._parse_state(line)
+
+    def reset(self) -> State:
+        self.serial.write('state reset')
+        line = self.serial.readline()
```

Comment:
Hmmm, in this case we what blocking read... homing may take about 20-30 seconds... @AndBondStyle

Reply:
In Marlin (and other g-code software, probably), timeouts are prevented by periodically sending a message like `:echo:busy:processing`. We could probably do the same. G28 is also homing command, btw.
![image](https://user-images.githubusercontent.com/18560701/128323568-8ceb0af7-f3f6-498d-8caa-aa64eccb2b09.png)

Reply:
Added 'echo:busy:processing' logic and moved it to SerialConnection class.